lupa/_lupa.pyx

```diff
@@ -1609,6 +1642,40 @@ cdef call_lua(LuaRuntime runtime, lua_State *L, tuple args):
     push_lua_arguments(runtime, L, args)
     return execute_lua_call(runtime, L, len(args))
 
+# adapted from https://stackoverflow.com/a/9672205
+cdef void* _lua_alloc_restricted(void* ud, void* ptr, size_t osize, size_t nsize) nogil:
+    cdef size_t* left = <size_t*>ud
+
+    if ptr is NULL:
+        # <http://www.lua.org/manual/5.2/manual.html#lua_Alloc>:
+        # When ptr is NULL, osize encodes the kind of object that Lua is
+        # allocating.
+        # Since we don’t care about that, just mark it as 0.
+        osize = 0
+
+    if nsize == 0:
+        free(ptr)
+        left[0] += osize # add old size to available memory
+        return NULL
+    elif nsize == osize:
+        return ptr
+    else:
+        if nsize > osize and left[0] < nsize - osize: # reached the limit
+            return NULL
+        ptr = realloc(ptr, nsize)
+        if ptr: # reallocation successful?
+            left[0] -= nsize + osize
+        return ptr
+
+cdef int _lua_panic(lua_State *L) nogil:
+    cdef char* msg = lua.lua_tostring(L, -1)
+    if msg == NULL:
+        msg = "error object is not a string"
+    cdef char* message = "PANIC: unprotected error in call to Lua API (%s)\n"
+    fprintf(stderr, message, msg)
+    fflush(stderr)
+    return 0 # return to Lua to abort
```

Comment:
I admit that jumping to address `1` (as done before) isn't exactly a great error handler, but it feels like we should generally improve the way fatal errors are handled, rather than making them fatal but just in other ways.

OTOH, this is an improvement all by itself, so… shouldn't it be a separate PR?

Reply:
This error handler is the one that luaL_newstate registers